### ISSUE TITLE
Specify model as explicit TightBindingModel type ?

### ISSUE THREAD

**Issue by jarvist** [jarvist/TheDancer.jl#3]
I've thought a little about the current `global const` nest of variables. 

I think the best way to deal with this may be to pass around an explicit `TightBindingModel` type.

Something like:
```julia
struct TightBindingModel
           N::Int
           dipoles
           psi
end
``` 

We can then have a direct constructor function to replace `init!` and `prepare_model`, and pass around the model as `m`, to operate on it:

```
unitary_propagation!(m::TightBindingModel, t)
    m.psi = 
``` 
etc.

I think that will make it more streamline + sensible, and the functions clean (without side-effects). What do you think @RoseTeague ?

**Comment by jarvist** [comment]
Started work on this with: https://github.com/jarvist/TheDancer.jl/commit/1228c4982dcaf8b353050e381c33166dbea25c9b

Got a bit bogged down in the Array view not passing over into the type!

**Comment by roseteague** [comment]
Hi Jarv, 
Sorry for not getting back to you on this sooner, I was stuck in bed yesterday feeling unwell. I think that sounds like a nice way of organising everything, I'm having a few issues with the global const form when I try and redefine the system parameters in the terminal which I think this would eliminate! 

Would this TightBindingModel include all the constants currently defined in init! (other than temperature and dipolestrength).

**Comment by jarvist** [comment]
Yes, I think all simulation-dependent parameters should go into the model spec, and can be carried around in a Struct Type

Then just the actual physical constants are in the global scope (i.e. hbar etc.), and you can cleanly setup the type constructor so that you fully specify the model as arguments.

You should then be able to do simultaneous and different simulations, i.e. something like
```
model1=TightBindingModel(N=20, dipolestrength=0.2)
model2=TightBindingModel(N=50, dipolestrength=0.1)
```

etc. without them stomping on each other.

**Comment by jarvist** [comment]
OK; so this is sort of ready to start implementing - it's going to require a lot of churn in the code which is presently there. 
As such, it probably makes sense to think a bit first about how to structure everything - and what the code interfaces should look like, with a view to the future and surface hopping / etc.

**Comment by roseteague** [comment]
Excellent! I will have a think about this today, but will focus on some other things first. I reckon the best thing to do will be to have a proper discussion on Monday about the structure and see what we agree/disagree on.

**Comment by jarvist** [comment]
I think that's a good idea Rose. I'll try and have a think, and do some pen-and-paper design of data structures etc. 
The other thing is that give implementing surface hopping is the next major goal (issue #2), we need data structures and a form that makes this easy. And in order to know what we need, we need to first understanding surface hopping!
So I'll try and read the Tully papers properly, and see how Schatz and Ratner describe it in overview. The main thing missing I think is the finite-difference estimate of the non-adiabatic coupling operator, for which you need to retain some (-dt) history of Psi.

**Comment by jarvist** [comment]
OK @RoseTeague , after our supervision meeting yesterday I've created a feature branch where I'll play with nicey-fying things in Julia while you are concentrating on the physics in here. If and when I get anywhere, I generate a formal pull-request and we can chat about stuff.
https://github.com/jarvist/TheDancer.jl/tree/jarv
